### PR TITLE
Adding option of timestamp to incrby. Change return value to [timestamp, value]

### PR DESCRIPTION
**Pull request by ashtul**
Issue #72

### CODE REVIEW

**Comment by gkorland** [comment]
@ashtul we need to fix the clients

**Comment by ashtul** [comment]
Feature requested in issue #72

**Comment by ashtul** [comment]
@danni-m @gkorland @otnielvh 
A small addition at API. Ability to give timestamp to INCRBY DECRBY commands.
Return value was left as OK to avoid breaking the API.

File: src/tests/test_module.py [hunk 1]
```diff
@@ -519,6 +519,18 @@ def test_incrby(self):
             assert result[0][0] >= start_incr_time
             assert len(result) <= 40
 
+    def test_incrby_with_timestamp(self):
+        with self.redis() as r:
+            r.execute_command('ts.create', 'tester')
+
+            for i in range(20):
+                assert r.execute_command('ts.incrby', 'tester', '5', 'TIMESTAMP', i) == 'OK'
```
Comment (gkorland) [inline src/tests/test_module.py]: @ashtul can you please also add test with RESET and batch events in the same reset time-backet?

Comment (ashtul) [reply src/tests/test_module.py]: Done

**Comment by ashtul** [comment]
@danni-m @gkorland API change removed. Now just add to the API.
Please review

**Comment by ashtul** [comment]
@danni-m @gkorland 
Solved conflict

File: src/module.c [hunk 2]
```diff
@@ -873,6 +873,20 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
                 series->lastValue = 0;
             }
         }
+        currentUpdatedTime = max(currentUpdatedTime, series->lastTimestamp);
+    }
+
+    long long timestamp = -1;
+    int timestampLoc = RMUtil_ArgIndex("TIMESTAMP", argv, argc);
```
Comment (gkorland) [inline src/module.c]: @ashtul should probably moved before line 863

File: src/module.c [hunk 3]
```diff
@@ -858,9 +858,24 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
         return RedisModule_WrongArity(ctx);
 
+
+    long long currentUpdatedTime = -1;
+    int timestampLoc = RMUtil_ArgIndex("TIMESTAMP", argv, argc);
+    if (timestampLoc > 0) {
+        if ((RedisModule_StringToLongLong(argv[timestampLoc + 1], (long long int *) &currentUpdatedTime) != REDISMODULE_OK)) {
+            // if timestamp is "*", take current time (automatic timestamp)
+            if(RMUtil_StringEqualsC(argv[timestampLoc + 1], "*")) {
+                currentUpdatedTime = (u_int64_t) RedisModule_Milliseconds();
```
Comment (gkorland) [inline src/module.c]: @ashtul  why do you need cast here? and not in line 874?

File: src/module.c [hunk 4]
```diff
@@ -858,9 +858,24 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     if (RMUtil_ParseArgs(argv, argc, 2, "d", &incrby) != REDISMODULE_OK)
         return RedisModule_WrongArity(ctx);
 
+
+    long long currentUpdatedTime = -1;
+    int timestampLoc = RMUtil_ArgIndex("TIMESTAMP", argv, argc);
+    if (timestampLoc > 0) {
+        if ((RedisModule_StringToLongLong(argv[timestampLoc + 1], (long long *) &currentUpdatedTime) != REDISMODULE_OK)) {
```
Comment (gkorland) [inline src/module.c]: @ashtul I think going positive it better
```C
if (timestampLoc > 0) {
        if ((RedisModule_StringToLongLong(argv[timestampLoc + 1], (long long *) &currentUpdatedTime) == REDISMODULE_OK)) {
            currentUpdatedTime = RedisModule_Milliseconds();
        } else if (RMUtil_StringEqualsC(argv[timestampLoc + 1], "*")) { 
            // if timestamp is "*", take current time (automatic timestamp)
            currentUpdatedTime = RedisModule_Milliseconds();
        } else {
            return RedisModule_ReplyWithError(ctx, "TSDB: invalid timestamp");
        }
} else {
        currentUpdatedTime = RedisModule_Milliseconds();
}
```

File: src/module.c [hunk 5]
```diff
@@ -873,6 +888,7 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
                 series->lastValue = 0;
             }
         }
+        currentUpdatedTime = max(currentUpdatedTime, series->lastTimestamp);
```
Comment (gkorland) [inline src/module.c]: @ashtul why did you move the max to the RESET block?

Comment (ashtul) [reply src/module.c]: Because it is relevant only if incrby has RESET flag.

File: src/module.c (hunk 4 above)
Comment (ashtul) [reply src/module.c]: @gkorland 
He never uses it this way in the code. I took as a code convention.
It is the same complexity and I agree your code is more elegant. It is your call.

Comment (gkorland) [reply src/module.c]: so please refactor it

File: src/module.c (hunk 5 above)
Comment (gkorland) [reply src/module.c]: @ashtul what was the effect of that not only on RESET?

Comment (ashtul) [reply src/module.c]: @gkorland the max wouldn't do anything since (except if you are using MacOS docker :) currentUpdatedTime would always be greater than series->lastTimestamp.
With RESET, it was normalized to resetMilliSeconds

File: src/module.c (hunk 4 above)
Comment (ashtul) [reply src/module.c]: Done

File: src/module.c [hunk 6]
```diff
@@ -858,9 +858,17 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     if (RMUtil_ParseArgs(argv, argc, 2, "d", &incrby) != REDISMODULE_OK)
         return RedisModule_WrongArity(ctx);
 
+
+    long long currentUpdatedTime = -1;
+    int timestampLoc = RMUtil_ArgIndex("TIMESTAMP", argv, argc);
+    if (timestampLoc == -1 || RMUtil_StringEqualsC(argv[timestampLoc + 1], "*")) {
```
Comment (danni-m) [inline src/module.c]: @ashtul @gkorland comment on this code block was correct, his code is easier to understand and will cause less confusion when looking at this parsing code.

Comment (gkorland) [reply src/module.c]: @danni-m it's ok, I think @ashtul is even cleaner than my
